doc: update installation and renew update chapters (#1329)
bareos/bareos: docs/manuals/source/IntroductionAndTutorial/UpdatingBareos.rst

.. _bareos-update:

Updating and upgrading Bareos
=============================

.. index::
   pair: Bareos; Update
   pair: Update; Linux

> .. _bareos_update_or_upgrade_preambule:

Reviewer: ```suggestion
.. _bareos_update_or_upgrade_preamble:
```

Before updating or upgrading Bareos software
--------------------------------------------

We consider both operations usually safe and can be applied without restriction to all installations.
We still would like to recommend to have done a number of security steps before going ahead.

> - Read the :ref:`Bareos current release notes <bareos-current-releasenotes>` to check all fixes done or any special instructions that can apply from your existing version to latest.

Reviewer: ```suggestion
- Read the :ref:`Bareos current release notes <bareos-current-releasenotes>`  and watch out for changes that  might impact your installation. Take special care for the :strong:`Breaking Changes` paragraph, as it contains the information about changes that require your special attention when upgrading your installation as they might require adaptions of the configuration.
```

- Update your operating system to its editor last security and patch level.
- Empty the running jobs queue.
- Run `BackupCatalog` or equivalent as last job (keep the most up to date database dump & configuration state).
- Stop all Bareos daemons.
- If not done before
   - Save a copy of the actual configuration used. (Usually :file:`/etc/bareos/` and :file:`/etc/bareos-webui/`.
   - Dump your database content to an easy to restore format.
- Cleanup the working dir :file:`/var/lib/bareos`

>    - Remove any old crash traces :file:`*backtrace` and :file:`*tracebac` files.

Author: I've fixed them, by using `*.bactrace` and `*.traceback`

   - Remove any left debugging :file:`*.trace` files.
   - Remove any leftover :file:`*.mail` that have not been sent.
   - Remove any :file:`*core*` that are left behind.
- Rotate log files if you don't use :command:`logrotate`
   - :file:`/var/log/bareos/bareos.log`
   - :file:`/var/log/bareos/bareos-audit.log`

Now you can safely apply the update or upgrade.

.. warning::

   Please remind, that |dir| and |sd| must always have the same version.
   The version of the File Daemon may differ, see chapter about :ref:`backward compatibility <backward-compatibility>`.

.. note::

   When you change the repository and refresh it, it is likely to have a new signing key.
   You will have to accept it.

.. code-block:: shell-session
   :caption: new gpg key detection on EL 8

>       Backup Archiving Recovery Open Sourced (EL_8)                                                                                                                                                               1.6 MB/s | 1.6 kB     00:00

Reviewer: ```suggestion
      Backup Archiving Recovery Open Sourced (EL_8)                                                           1.6 MB/s | 1.6 kB     00:00
```

      Key Algorithm:    RSA 4096
      Key Created:      Mon Dec 20 10:04:50 2021
      Key Expires:      (does not expire)
      Rpm Name:         gpg-pubkey-c9fed482-61c05542

         Note: Signing data enables the recipient to verify that no modifications occurred after the data
         were signed. Accepting data with no, wrong or unknown signature can lead to a corrupted system
         and in extreme cases even to a system compromise.

         Note: A GPG pubkey is clearly identified by its fingerprint. Do not rely on the key's name. If
         you are not sure whether the presented key is authentic, ask the repository provider or check
         their web site. Many providers maintain a web page showing the fingerprints of the GPG keys they
         are using.

      Do you want to reject the key, trust temporarily, or trust always? [r/t/a/?] (r): a```

.. _bareos-update_from_org_to_com:

Updating from community to subscription binaries
------------------------------------------------

To update the installed packages to Bareos Subscription binaries, you will have to point to the subscription repositories located at https://download.bareos.com/.

> Once you received your `download.bareos.com` portal credentials, you can refer to the following section :ref:`section-AddSoftwareRepository` for complete instructions how-to use the :file:`add_bareos_repositories.sh` helper.

Reviewer: ```suggestion
Once you received your `download.bareos.com` portal/repository credentials, you can refer to the following section :ref:`section-AddSoftwareRepository` for complete instructions how-to use the :file:`add_bareos_repositories.sh` helper.
```

> Choose the same operating system and Bareos major version you are already using. You will then benefit the last bug fixes and enhancements, without having to do a major upgrade immediately.

Reviewer: ```suggestion
Choose the same operating system and Bareos major version you are already using.
```
Strictly speaking, this is not true anymore after we have `current` on the bareos.org server.

Author: I agree, but then maybe we will speak about upgrading to current for community. Here the chapter can be applied to both and for d.b.c it will still be the case.

Read the :ref:`Bareos current release notes <bareos-current-releasenotes>` to check all fixes that have been made.

.. note::

   You can install directly the latest Major,Minor,Bugfix release available.
   So updating from 21.0.0 directly to 21.1.5 is not a problem.

.. _bareos-update_minor_bugfix:

.. index::
   pair: Bareos; Update
   pair: Update; Linux
   pair: update; minor;
   pair: update; bugfix

Updating Bareos to last minor bugfix release
--------------------------------------------

In most cases, a Bareos update is simply done by a package update of the distribution.

.. note::

>    Please before processing, apply steps in :ref:`bareos_update_or_upgrade_preambule`

Reviewer: ```suggestion
   Please before processing, apply steps in :ref:`bareos_update_or_upgrade_preamble`
```

In this section, we explain how to update your Bareos major version to the latest minor,bugfix release.

For upgrading to a new major version see :ref:`Update Bareos to a new major release <bareos-upgrade_major>`.

Example how to update from 21.0.0 to 21.1.5.

.. index::
   single: update minor bugfix; RHEL
   single: update minor bugfix; CentOS
   single: update minor bugfix; Fedora
   single: update minor bugfix; EL

.. code-block:: shell-session
   :caption: Shell example command to update Bareos on on EL 8

   root@host:~# dnf upgrade --repo=bareos --refresh

>       Backup Archiving Recovery Open Sourced (EL_8)                                                                                                                                                               1.5 kB/s | 833  B     00:00

Reviewer: ```suggestion
      Backup Archiving Recovery Open Sourced (EL_8)                                                              1.5 kB/s | 833  B     00:00
```

      bareos-database-postgresql      x86_64           21.1.5-3.el8    bareos       42 k
      bareos-database-tools           x86_64           21.1.5-3.el8    bareos      107 k
      bareos-director                 x86_64           21.1.5-3.el8    bareos      425 k
      bareos-filedaemon               x86_64           21.1.5-3.el8    bareos      120 k
      bareos-storage                  x86_64           21.1.5-3.el8    bareos       97 k
      bareos-tools                    x86_64           21.1.5-3.el8    bareos       52 k

      Transaction Summary
      ====================================================================================
      Upgrade  11 Packages

      Total download size: 1.7 M
      Is this ok [y/N]: y

.. index::
   single: update minor bugfix; SLE
   single: update minor bugfix; openSUSE

.. code-block:: shell-session
   :caption: Shell example command to update Bareos on SLES / openSUSE

   root@host:~# zypper refresh --force bareos
   root@host:~# zypper -v update --repo=bareos
      Verbosity: 2
      Initializing Target
      Checking whether to refresh metadata for bareos

>       Retrieving: repomd.xml ...................................................................................................................................................................................................[done (3.0 KiB/s)]
>       Retrieving: media ...............................................................................................................................................................................................................[not found]
>       Retrieving: repomd.xml.asc ...........................................................................................................................................................................................................[done]
>       Retrieving: repomd.xml.key ...........................................................................................................................................................................................................[done]
>       Retrieving: repomd.xml ...............................................................................................................................................................................................................[done]
>       Repository:       bareos
>       Key Fingerprint:  91DA 1DC3 564A E20A 76C4 CA88 E019 57D6 C9FE D482
>       Key Name:         Bareos 21 Signing Key <signing@bareos.com>
>       Key Algorithm:    RSA 4096
>       Key Created:      Mon Dec 20 10:04:50 2021
>       Key Expires:      (does not expire)
>       Rpm Name:         gpg-pubkey-c9fed482-61c05542
>       Retrieving: 7c2078b9b802f0f5c4edb818e870be0084ae132b4a5f21111617582fd927a65f-primary.xml.gz ..........................................................................................................................................[done]
>       Retrieving repository 'bareos' metadata ..............................................................................................................................................................................................[done]
>       Building repository 'bareos' cache ...................................................................................................................................................................................................[done]
>       Loading repository data...

Reviewer: ```suggestion
      Retrieving: repomd.xml ...................................................................................................................[done (3.0 KiB/s)]
      Retrieving: media .............................................................................................................................[not found]
      Retrieving: repomd.xml.asc ............................................................................................................[done]
      Retrieving: repomd.xml.key ............................................................................................................[done]
      Retrieving: repomd.xml ...................................................................................................................[done]
      Repository:       bareos
      Key Fingerprint:  91DA 1DC3 564A E20A 76C4 CA88 E019 57D6 C9FE D482
      Key Name:         Bareos 21 Signing Key <signing@bareos.com>
      Key Algorithm:    RSA 4096
      Key Created:      Mon Dec 20 10:04:50 2021
      Key Expires:      (does not expire)
      Rpm Name:         gpg-pubkey-c9fed482-61c05542
      Retrieving: 7c2078b9b802f0f5c4edb818e870be0084ae132b4a5f21111617582fd927a65f-primary.xml.gz ...[done]
      Retrieving repository 'bareos' metadata .....................................................................................[done]
      Building repository 'bareos' cache ...............................................................................................[done]
      Loading repository data...
```

   :caption: Shell example command to check bareos configuration

   root@host:~# bareos-fd -t -u root -g bareos
   root@host:~# bareos-sd -t -u bareos -g bareos
   root@host:~# bareos-dir -t -u bareos -g bareos

Depending of the operating system and its configuration, you will have to restart the daemons.
Use your operating system command to do so.

.. code-block:: shell-session
   :caption: Shell command to restart all bareos daemon with systemd on Linux

   root@host:~# systemctl restart bareos-director bareos-storage bareos-filedaemon
   root@host:~# systemctl status bareos-director bareos-storage bareos-filedaemon

.. code-block:: shell-session
   :caption: Shell command to restart all bareos daemon with service on FreeBSD

   root@host:~# service bareos-dir restart
   root@host:~# service bareos-fd restart
   root@host:~# service bareos-sd restart

.. _bareos-upgrade_major:

.. index::
   pair: Bareos; Upgrade
   pair: Update; Major

Upgrading Bareos to a new major release
---------------------------------------

In most cases, a Bareos major upgrade can be achieve by:

- Add new major directory (subscription only)
- Package upgrade of the distribution.

> - Database schema upgrade with helper scripts.

Reviewer: ```suggestion
- Database schema upgrade with helper scripts (if schema was changed).
```

- Configuration review to cleanup deprecated or removed parameters.
- Review of home made scripts and manage their adaptation in case of changes.

Prepare the upgrade
~~~~~~~~~~~~~~~~~~~

> If you not have already done those steps, please refer to instructions in :ref:`bareos_update_or_upgrade_preambule`.

Reviewer: ```suggestion
If you not have already done those steps, please refer to instructions in :ref:`bareos_update_or_upgrade_preamble`.
```

> .. warning:

Reviewer: ```suggestion
.. warning::
```
This typo makes this warning be a comment.

>    If you use any third party plugins, you should check and test their functionalities with the new major version before hand.

Reviewer: ```suggestion
   If you use any third party plugins, you should check and test their functionalities with the new major version beforehand.
```

Upgrade major directory on download repositories
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> - First remove the existing bareos repositories definitions, by either removing the definition file(s) or by using your package manager.

Reviewer: ```suggestion
- First remove the existing bareos repository definitions, by either removing the definition file(s) or by using your package manager.
```

- Point your browser to the new Bareos major version for your operating system on the download server.
- Open or save the helper script :file:`add_bareos_repositories.sh`.

>   + You can refer to the following section :ref:`section-AddSoftwareRepository` for complete instructions how-to use the :file:`add_bareos_repositories.sh` helper.

Reviewer: ```suggestion
  + You can refer to the following section :ref:`section-AddSoftwareRepository` for complete instructions how to use the :file:`add_bareos_repositories.sh` helper.
```

- Transfer the file to your bareos server, and execute it as root.
   new repositories definitions will be created.

.. code-block:: shell-session
   :caption: Shell command to upgrade Bareos Major version

   root@host:~# sh add_bareos_repositories.sh

You should be able now to proceed the appropriate commands to refresh the packages list and upgrade the package to the newer version.

> .. note:

Reviewer: ```suggestion
.. note::
```
This typo again makes this a comment.

   You can refer to section :ref:`bareos-update_minor_bugfix` for commands example.

Updating the configuration files
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> When updating Bareos through the distribution packaging mechanism, the existing configuration kept as they are.

Reviewer: ```suggestion
When updating Bareos through the distribution packaging mechanism, the existing configuration files are kept as they are.
```

If you don’t want to modify the behavior, there is normally no need to modify the configuration.

However, in some rare cases, configuration changes are required. These cases are described in the :ref:`Release Notes <releasenotes>`.

> It is generally sufficient, to upgrade to the latest release, without having to install any intermediate releases

Reviewer: ```suggestion
It is generally sufficient  to upgrade to the latest release, without having to install any intermediate releases
```

Author: double space I will fix it.

         are using.

      Do you want to reject the key, trust temporarily, or trust always? [r/t/a/?] (r): a

.. _section-UpdateFromCommunityToSubscription:

Updating from community to subscription binaries
------------------------------------------------

.. index::
    single: update from community to subscription
    pair: Bareos; community repository
    pair: Bareos; subscription repository

To update the installed community packages (https://download.bareos.org) to Bareos Subscription packages,
you will have to point to the subscription repositories located at https://download.bareos.com/.

Once you received your `download.bareos.com` portal/repository credentials, you can refer to the following section :ref:`section-AddSoftwareRepository` for complete instructions how-to use the :file:`add_bareos_repositories.sh` helper.

Choose the same operating system and Bareos major version you are already using.

Read the :ref:`Bareos current release notes <bareos-current-releasenotes>` to check all fixes that have been made.

Proceed to the next section, to install last minor bugfix release on your systems.

.. _section-UpdateMinorBugfix:

> Updating Bareos to lastest minor or bugfix release

Reviewer 2: ```suggestion
Updating Bareos to the lastest minor or bugfix release
```

      1 MiB to be downloaded.
      Proceed with this action? [y/N]: y

.. _section-UpdatePostChecks:

Post update checks
~~~~~~~~~~~~~~~~~~

.. index::
   single: Update post checks

After the update, it is recommended to check if any new warnings are raised when starting the daemon, mostly deprecated configuration directives.
Bareos will mark configuration directives at least for one major release as deprecated, before removing them.

To do so you can use the `-t` flag:

.. code-block:: shell-session
   :caption: Shell example to check the Bareos configuration

   root@host:~# su - bareos -s /bin/sh -c "bareos-dir -t"
   root@host:~# su - bareos -s /bin/sh -c "bareos-sd -t"
   root@host:~# bareos-fd -t
   There are configuration warnings:
      * using deprecated keyword PidDirectory on line 19 of file /etc/bareos/bareos-fd.d/client/myself.conf

The same warnings are also shown on a regular start of the daemons.

Depending of the operating system and its configuration, you will have to restart the daemons.
Use your operating system command to do so.

.. code-block:: shell-session
   :caption: Shell command to restart all bareos daemon with systemd on Linux

>    root@host:~# systemctl restart bareos-director bareos-storage bareos-filedaemon
>    root@host:~# systemctl status bareos-director bareos-storage bareos-filedaemon

Reviewer 2: ```suggestion
   root@host:~# systemctl restart bareos-director.service bareos-storage.service bareos-filedaemon.service
   root@host:~# systemctl status bareos-director.service bareos-storage.service bareos-filedaemon.service
```
Isn't it recommended to specify services with the service extension?

Author: Never used the .service you don't need it, except when you have several different units with the same name like a .service and .timer or .target for example.

- Read the :ref:`Bareos current release notes <bareos-current-releasenotes>`  and watch out for changes that  might impact your installation. Take special care for the :strong:`Breaking Changes` paragraph, as it contains the information about changes that require your special attention when upgrading your installation as they might require adaptions of the configuration.
- Check for deprecated configuration directives.

   - Bareos will mark configuration directives at least for one major release as deprecated, before removing them.

      .. code-block:: shell-session
         :caption: Shell example to check the Bareos configuration

         root@host:~# su - bareos -s /bin/sh -c "bareos-dir -t"
         root@host:~# su - bareos -s /bin/sh -c "bareos-sd -t"
         root@host:~# bareos-fd -t
         There are configuration warnings:
          * using deprecated keyword PidDirectory on line 19 of file /etc/bareos/bareos-fd.d/client/myself.conf

   - The same warnings are also shown on a regular start of the daemons.

- Update your operating system to the latest security and patch level of the publisher.
- Empty the running jobs queue.

> - Run :strong:`BackupCatalog` or equivalent as last job (keep the most up to date database dump & configuration state).
> - Stop all Bareos daemons.
> - If not done before
> 
>    - Save a copy of the actual configuration used. (Usually :file:`/etc/bareos/` and :file:`/etc/bareos-webui/`.
>    - Dump your database content to an easy to restore format.

Reviewer 2: ```suggestion
- Stop all Bareos daemons.
- Run :strong:`BackupCatalog` or equivalent as last job (keep the most up to date database dump & configuration state).

   - or

      - Save a copy of the actual configuration used. (Usually :file:`/etc/bareos/` and :file:`/etc/bareos-webui/`.
      - Dump your database content to an easy to restore format.
```

How about this? Either make a proper Database backup by the existing job or manually (but not twice).

Author: This can't work, because you will not be able to run backupcatalog with stopped daemon.

In real life, if you don't tweak our dump script to stop erasing the db dump, you really want to have a ready to use dump instead of extracting the one included inside the volume with bextract. The delta can be hours.

      bareos-client                   x86_64           21.1.5-3.el8    bareos      7.5 k
      bareos-common                   x86_64           21.1.5-3.el8    bareos      764 k
      bareos-database-common          x86_64           21.1.5-3.el8    bareos       87 k
      bareos-database-postgresql      x86_64           21.1.5-3.el8    bareos       42 k
      bareos-database-tools           x86_64           21.1.5-3.el8    bareos      107 k
      bareos-director                 x86_64           21.1.5-3.el8    bareos      425 k
      bareos-filedaemon               x86_64           21.1.5-3.el8    bareos      120 k
      bareos-storage                  x86_64           21.1.5-3.el8    bareos       97 k
      bareos-tools                    x86_64           21.1.5-3.el8    bareos       52 k

      Transaction Summary
      ====================================================================================
      Upgrade  11 Packages

      Total download size: 1.7 M
      Is this ok [y/N]: y

.. index::
   single: update minor bugfix; SLE
   single: update minor bugfix; openSUSE

.. code-block:: shell-session
   :caption: Shell example command to update Bareos on SLES / openSUSE

>    root@host:~# zypper refresh --force bareos

Reviewer 2: Why `--force`? Force always leaves the impression, that you do something you shouldn'd do.

Author: because if the project publish something recently (between the lap of time the cache is valid) the metadata will not be in sync with the repository content. Maybe because I'm use to Tumbleweed and get often hit by this kind of things.

      Checking for upgrades (8 candidates): 100%
      Processing candidates (8 candidates): 100%
      The following 8 package(s) will be affected (of 0 checked):
      Installed packages to be UPGRADED:
            bareos.com-bconsole: 21.0.0 -> 21.1.5 [Bareos]
            bareos.com-common: 21.0.0 -> 21.1.5 [Bareos]
            bareos.com-database-common: 21.0.0 -> 21.1.5 [Bareos]
            bareos.com-database-postgresql: 21.0.0 -> 21.1.5 [Bareos]
            bareos.com-database-tools: 21.0.0 -> 21.1.5 [Bareos]
            bareos.com-director: 21.0.0 -> 21.1.5 [Bareos]
            bareos.com-filedaemon: 21.0.0 -> 21.1.5 [Bareos]
            bareos.com-storage: 21.0.0 -> 21.1.5 [Bareos]
      Number of packages to be upgraded: 8
      1 MiB to be downloaded.
      Proceed with this action? [y/N]: y

.. _section-UpdatePostChecks:

Post update checks
~~~~~~~~~~~~~~~~~~

.. index::
   single: Update post checks

After the update, it is recommended to check if any new warnings are raised when starting the daemon.

To do so you can use the `-t` flag:

.. code-block:: shell-session
   :caption: Shell example command to check bareos configuration

>    root@host:~# bareos-fd -t -u root -g bareos
>    root@host:~# bareos-sd -t -u bareos -g bareos
>    root@host:~# bareos-dir -t -u bareos -g bareos

Reviewer 2: ```suggestion
   root@host:~# su - bareos -s /bin/sh -c "bareos-dir -t"
   root@host:~# su - bareos -s /bin/sh -c "bareos-sd -t"
   root@host:~# bareos-fd -t
```
I always use these commands. Don't know, what is preferable.

Author: don't know which one it the best, especially to detect non accessible config file or working directory for example.

It is generally sufficient to upgrade directly to the latest release, without having to install any intermediate releases.
However, it is required to read the release notes of all intermediate releases.

One exception is when using a |mysql| Bareos catalog,
which have been removed with Bareos :sinceVersion:`21.0.0: MySQL backend removed`.
Therefore you first have to upgrade to Bareos 20 and migrate the |mysql| into a |postgresql| Bareos Catalog, see :ref:`section-MigrationMysqlToPostgresql`.

Prepare the upgrade
~~~~~~~~~~~~~~~~~~~

If you not have already done those steps, please refer to instructions in :ref:`section-UpdateUpgradePreamble`.

.. warning::

   If you use any third party plugins, you should check and test their functionalities with the new major version beforehand.

.. _sectionUgradeMajorRepository:

Upgrade the Bareos download repositories
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This does only apply for subscription repositories h(ttps://download.bareos.com/bareos/release/).
The community repository (https://download.bareos.org/current) will always contain the latest build of the most recent stable branch.

> - First remove the existing Bareos repository definitions, by either removing the definition file(s) or by using your package manager.

Reviewer 2: Is this really required? The package manage should detect the latest version anyway and the `add_bareos_repository.sh` script overwrites the existing repository anyway (on all platforms?).

Author: I've been hit at least on one platform, I will try to reproduce it again.

Author: ```
 bash add_bareos_repositories.sh
Adding repository 'bareos' .............................................................................................................................................................................[error]
Repository named 'bareos' already exists. Please use another alias.

See :ref:`section-UpdateToConfigurationSubdirectories` for details and how to migrate to :ref:`section-SubdirectoryConfigurationScheme`.

.. _sectionUpdateDatabaseScheme:

Updating the database scheme
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Sometimes improvements in Bareos make it necessary to update the database scheme.

.. warning::

   If the Bareos catalog database does not have the current schema, the Bareos Director refuses to start.

.. code-block:: shell-session
   :caption: Shell example of bareos-dir failing to start due to lack of database schema update

>    root@host:~# bareos-dir -t

Reviewer 2: ```suggestion
   root@host:~# su - bareos -s /bin/sh -c "bareos-dir -t"
```
The permission to the database bareos is normally only given to the system user bareos. So, be default, root has no permission to access this database.
Also running bareos-dir as root can result in creating files accessable only as root, which can cause problem by a later regular start (even if I think, this does not happen with the `-t` parameter)

Author: was forgotten, otherwise I would have appy my way with bareos-dir -t -u bareos -g bareos ;-)

The :command:`grant_bareos_privileges` command is required, if new databases tables are introduced. It does not hurt to run it multiple times.

After this, restart the Bareos Director and verify it starts without problems.

.. warning::

   If you're updating from Bareos <= 13.2.3 and have configured the Bareos database during install using Bareos environment variables
   (``db_name``\ , ``db_user``\  or ``db_password``\ , see :ref:`CatMaintenanceChapter`), make sure to have these variables defined
   in the same way when calling the update and grant scripts. Newer versions of Bareos read these variables from the Director configuration.
   However, make sure that the user running the database scripts has read access to this file (or set the environment variables).

>    The **postgres** user normally does not have the required permissions.

Reviewer: Can we drop this part? I think bareos 13 is really too old.

Author: I'm in favor of. 
Thanks for having done this review.

Author: Done.
